File: java/com/google/openlocationcode/tests/ShorteningTest.java
package com.google.openlocationcode.tests;

import static java.nio.charset.StandardCharsets.UTF_8;

import com.google.openlocationcode.OpenLocationCode;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Tests shortening functionality of Open Location Code. */
@RunWith(JUnit4.class)
public class ShorteningTest {

  private static class TestData {

    private final String code;
    private final double referenceLatitude;
    private final double referenceLongitude;
    private final String shortCode;
    private final String testType;

    public TestData(String line) {
      String[] parts = line.split(",");
      if (parts.length != 5) {
        throw new IllegalArgumentException("Wrong format of testing data.");
      }
      this.code = parts[0];
      this.referenceLatitude = Double.valueOf(parts[1]);
      this.referenceLongitude = Double.valueOf(parts[2]);
      this.shortCode = parts[3];
      this.testType = parts[4];
    }
  }

  private static final String TEST_PATH =
      System.getenv("JAVA_RUNFILES") + "/openlocationcode/test_data";
  private final List<TestData> testDataList = new ArrayList<>();

  @Before
  public void setUp() throws Exception {
    File testFile = new File(TEST_PATH, "shortCodeTests.csv");
    InputStream testDataStream = new FileInputStream(testFile);
    BufferedReader reader = new BufferedReader(new InputStreamReader(testDataStream, UTF_8));
    String line;
    while ((line = reader.readLine()) != null) {
      if (line.startsWith("#")) {
        continue;
      }
      testDataList.add(new TestData(line));
    }
  }

  @Test
  public void testShortening() {
    for (TestData testData : testDataList) {
      if (!"B".equals(testData.testType) && !"S".equals(testData.testType)) {
        continue;
      }
      OpenLocationCode olc = new OpenLocationCode(testData.code);
      OpenLocationCode shortened =
          olc.shorten(testData.referenceLatitude, testData.referenceLongitude);
      Assert.assertEquals(
          "Wrong shortening of code " + testData.code,
          testData.shortCode,
          shortened.getCode());
    }
  }

  @Test
  public void testRecovering() {
    for (TestData testData : testDataList) {
      if (!"B".equals(testData.testType) && !"R".equals(testData.testType)) {
        continue;
      }
      OpenLocationCode olc = new OpenLocationCode(testData.shortCode);
      OpenLocationCode recovered =
          olc.recover(testData.referenceLatitude, testData.referenceLongitude);
      Assert.assertEquals(testData.code, recovered.getCode());
    }
  }
}

File: kotlin/src/commonTest/kotlin/com/google/openlocationcode/ShorteningTest.kt
package com.google.openlocationcode

import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

class ShorteningTest {
    data class TestData(
        val code: String,
        val referenceLatitude: Double,
        val referenceLongitude: Double,
        val shortCode: String,
        val testType: String
    ) {
        companion object {
            fun fromLine(line: String): TestData {
                val parts = line.split(',')
                require(parts.size == 5) { "Wrong format of testing data." }
                return TestData(
                    code = parts[0],
                    referenceLatitude = parts[1].toDouble(),
                    referenceLongitude = parts[2].toDouble(),
                    shortCode = parts[3],
                    testType = parts[4]
                )
            }
        }
    }


    private lateinit var testDataList: List<TestData>

    @BeforeTest
    fun setup() {
        testDataList =
            loadTestCsvFileAsLines("shortCodeTests.csv")
                .filterNot { it.startsWith('#') }
                .map { TestData.fromLine(it) }
    }


    @Test
    fun testShortening() {
        for (testData in testDataList) {
            if ("B" != testData.testType && "S" != testData.testType) {
                continue
            }
            val olc = OpenLocationCode(testData.code)
            val shortened = olc.shorten(testData.referenceLatitude, testData.referenceLongitude)
            assertEquals(testData.shortCode, shortened.code, "Wrong shortening of code ${testData.code}")
        }
    }

    @Test
    fun testRecovering() {
        for (testData in testDataList) {
            if ("B" != testData.testType && "R" != testData.testType) {
                continue
            }
            val olc = OpenLocationCode(testData.shortCode)
            val recovered = olc.recover(testData.referenceLatitude, testData.referenceLongitude)
            assertEquals(testData.code, recovered.code)
        }
    }
}
